First draft of new README.md (#1182)
GoogleCloudPlatform/magic-modules: README.md

## Overview

Magic Modules is a tool we use to autogenerate infrastructure-as-code tools for
Google Cloud Platform. GCP ["resources"](https://cloud.google.com/docs/overview/#gcp_resources)
are encoded in a shared data file, and that data is used to fill in
"Mad Libs"-style templates across each of the tools Magic Modules generates.

They include:

* Terraform
* Ansible
* InSpec

Not only is Magic Modules a force multiplier for our developers, Magic Modules
allows us to preemptively solve issues by encoding field-tested learnings about
the GCP API into each tool; issues solved for one tool will be fixed in each
other tool.

## Getting Started with Magic Modules

> Using Magic Modules to make changes to the generated tools might seem

Reviewer: I'd get rid of this paragraph all-together. I think it might unintentionally scare off contributors. After all, thanks to new documentation, it shouldn't be that intimidating at all!

intimidating, but we're here to help you get familiar with our tooling.

We've prepared an interactive tutorial that you can try out with Open in Cloud
Shell below; if you're getting set up on a local workstation, this guide serves
as a reference.

[![Open in Cloud Shell](http://gstatic.com/cloudssh/images/open-btn.svg)](https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/magic-modules&tutorial=TUTORIAL.md)

### Requirements

To get started, you'll need:

> * Ruby 2.5.0

Reviewer: ```suggestion
* Ruby 2.5.0+
```

Author: I got a warning when using 2.5.1 so let's leave this as-is for now.

>   * You can use `rbenv` to manage your Ruby version(s)

Reviewer: ```suggestion
  * (optional): You can use `rbenv` or `rvm` to install + manage your Ruby installation.
```

Author: This is already worded like a suggestion so no need to add optional imo. Also suggesting multiple tools just makes things complicated, let's keep it simple. (I'm on the fence about including this blurb altogether tbh- we shouldn't worry about how users have configured the language on their system)

> * The [`bundler`](https://github.com/bundler/bundler) gem

Reviewer: ```suggestion
* [`Bundler`](https://github.com/bundler/bundler)
```

  * This can be installed with `gem install bundler`

### Preparing Magic Modules / One-time setup

> If you're developing Ansible or Inspec, we use submodules to manage the Magic

Reviewer: I'm a little worried that the README is overlapping too much with the tutorial with some of this setup info.

Author: I don't think the tutorial should be the exclusive means of learning how to use MM.

ansible   | build/ansible
inspec    | build/inspec

For example, to generate Terraform:

```
bundle exec compiler -a -v "ga" -e terraform -o "$GOPATH/src/github.com/terraform-providers/terraform-provider-google"
```

It's worth noting that Magic Modules will only generate new files when ran
locally. The "Magician"- the Magic Modules CI system- handles deletion of old
files when creating PRs.

#### Terraform's `google-beta` provider

Terraform is the only tool to handle Beta features right now; you can generate
`google-beta` by running the following, substitution `"beta"` for the version
and using the repository for the `google-beta` provider.

```
bundle exec compiler -a -v "beta" -e terraform -o "$GOPATH/src/github.com/terraform-providers/terraform-provider-google-beta"
```

### Testing your changes

Once you've generated your changes for the tool, you can test them by running the
tool-specific tests as if you were submitting a PR against that tool.

You can run tests in the `{{output_folder}}` from above. See the following for
more details;

Tool             | Testing Guide
-----------------|--------------

> ansible          | [instructions](https://docs.ansible.com/ansible/devel/dev_guide/testing.html)

Reviewer: I feel like we might want to put our own instructions, but this is a good starting point.

Author: My intention here is that we just defer to the downstreams; users should read how to test Ansible, or Inspec, or Terraform, and not how to use MM to test them.

Author: I'll leave it to you and @slevenick to fill in whatever you want for your tools afterwards, but at least for Terraform the entire story right now is "test it like it's handwritten code"
